{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import datasets\n",
    "from scipy.optimize import minimize\n",
    "from scipy.special import expit\n",
    "from sklearn.metrics import log_loss\n",
    "from softmax import softmax, log_softmax\n",
    "import theano\n",
    "from theano import tensor as T\n",
    "import tensorflow as tf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "X, y = datasets.htwt()\n",
    "Y = datasets.one_hot(y)\n",
    "N, D = X.shape\n",
    "N, C = Y.shape\n",
    "Xaug = np.hstack([np.ones((N, 1)), X])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR (theano.gof.opt): Optimization failure due to: constant_folding\n",
      "ERROR:theano.gof.opt:Optimization failure due to: constant_folding\n",
      "ERROR (theano.gof.opt): TRACEBACK:\n",
      "ERROR:theano.gof.opt:TRACEBACK:\n",
      "ERROR (theano.gof.opt): Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n",
      "ERROR:theano.gof.opt:Traceback (most recent call last):\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/opt.py\", line 1762, in process_node\n",
      "    replacements = lopt.transform(node)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/opt.py\", line 5445, in constant_folding\n",
      "    required = thunk()\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/gof/op.py\", line 865, in rval\n",
      "    r = p(n, [x[0] for x in i], o)\n",
      "  File \"/home/greg/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py\", line 5187, in perform\n",
      "    z[0] = numpy.asarray(numpy.dot(x, y))\n",
      "ValueError: shapes (210,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "z = T.dvector()\n",
    "mu = lambda X, W, b: T.nnet.sigmoid(T.dot(X, w) + b)\n",
    "loss = T.sum(T.nnet.binary_crossentropy(mu(X, *decode(z)), Y))\n",
    "grad = theano.function([z], T.grad(loss, z))\n",
    "hess = theano.function([z], T.hessian(loss, z))    \n",
    "loss = theano.function([z], loss)\n",
    "decode = lambda params: (params[:D], params[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "shapes (6,) and (3,) not aligned: 6 (dim 0) != 3 (dim 0)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-209-ab05740a3d9f>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      7\u001b[0m \u001b[0mparams\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m*\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mD\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      8\u001b[0m \u001b[1;31m# w = minimize(loss, params).x\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 9\u001b[1;33m \u001b[0mw\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mminimize\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mloss\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mparams\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mmethod\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;34m'Newton-CG'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mjac\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mgrad\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mhess\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mhess\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     10\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     11\u001b[0m \u001b[0mgrad\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mw\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/home/greg/anaconda/lib/python2.7/site-packages/scipy/optimize/_minimize.pyc\u001b[0m in \u001b[0;36mminimize\u001b[1;34m(fun, x0, args, method, jac, hess, hessp, bounds, constraints, tol, callback, options)\u001b[0m\n\u001b[0;32m    420\u001b[0m     \u001b[1;32melif\u001b[0m \u001b[0mmeth\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;34m'newton-cg'\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    421\u001b[0m         return _minimize_newtoncg(fun, x0, args, jac, hess, hessp, callback,\n\u001b[1;32m--> 422\u001b[1;33m                                   **options)\n\u001b[0m\u001b[0;32m    423\u001b[0m     \u001b[1;32melif\u001b[0m \u001b[0mmeth\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;34m'anneal'\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    424\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0m_minimize_anneal\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfun\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mx0\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0margs\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0moptions\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/home/greg/anaconda/lib/python2.7/site-packages/scipy/optimize/optimize.pyc\u001b[0m in \u001b[0;36m_minimize_newtoncg\u001b[1;34m(fun, x0, args, jac, hess, hessp, callback, xtol, eps, maxiter, disp, return_all, **unknown_options)\u001b[0m\n\u001b[0;32m   1403\u001b[0m                     \u001b[0mhcalls\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mhcalls\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1404\u001b[0m             \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1405\u001b[1;33m                 \u001b[0mAp\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mnumpy\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdot\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mA\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mpsupi\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1406\u001b[0m             \u001b[1;31m# check curvature\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1407\u001b[0m             \u001b[0mAp\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0masarray\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mAp\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msqueeze\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m  \u001b[1;31m# get rid of matrices...\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mValueError\u001b[0m: shapes (6,) and (3,) not aligned: 6 (dim 0) != 3 (dim 0)"
     ]
    }
   ],
   "source": [
    "mu = lambda X, w: expit(Xaug.dot(w))\n",
    "loss = lambda w: log_loss(y, mu(X, w))\n",
    "grad = lambda w: Xaug.T.dot(mu(X, w) - y)\n",
    "def hess(w):\n",
    "    a = mu(X, w)\n",
    "    return X.T.dot(np.diag(a * (1 - a))).dot(Xaug).ravel()\n",
    "\n",
    "params = [0] * (D + 1)\n",
    "# w = minimize(loss, params).x\n",
    "w = minimize(loss, params, method = 'Newton-CG', jac = grad, hess = hess).x\n",
    "\n",
    "grad(w)\n",
    "hess(w)\n",
    "yhat = mu(X, w) > 0.5\n",
    "print '%0.3f' % np.mean(yhat != y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.348\n"
     ]
    }
   ],
   "source": [
    "mu = lambda X, W, b: softmax(X.dot(W) + b)\n",
    "\n",
    "loss = lambda params: log_loss(y, mu(X, *decode(params)))\n",
    "\n",
    "def grad(params):\n",
    "    Z = mu(X, *decode(params)) - Y\n",
    "    dW = sum([np.kron(z, X[i]) for i, z in enumerate(Z)])\n",
    "    dW = np.array(dW.reshape((C, C))).T\n",
    "    db = np.sum(Z, axis = 0)\n",
    "    return np.array(list(dW.ravel()) + list(db.ravel()))\n",
    "\n",
    "def hess(params):\n",
    "    Z = mu(X, *decode(params))\n",
    "    o = lambda x: np.outer(x, x)\n",
    "    dW = sum([np.kron(np.diag(z) - o(z), o(X[i])) for i, z in enumerate(Z)])\n",
    "    db = sum([np.diag(z) - o(z) for i, z in enumerate(Z)])\n",
    "    return np.array([0] * 36).reshape((6, 6))\n",
    "    return np.array(list(dW.ravel()) + list(db.ravel()))\n",
    "\n",
    "params = [0] * (D * (C - 1) + C)\n",
    "decode = lambda params: (np.array((list(params[:D * (C - 1)]) + [0] * C)).reshape((D, C)), params[-C:])\n",
    "\n",
    "params = [0] * (D * C + C)\n",
    "decode = lambda params: (np.array((list(params[:D * C]))).reshape((D, C)), params[-C:])\n",
    "\n",
    "# params = minimize(loss, params).x\n",
    "params = minimize(loss, params, method = 'Newton-CG', jac = grad, hess = hess).x\n",
    "yhat = np.argmax(mu(X, *decode(params)), axis = 1)\n",
    "print '%0.3f' % np.mean(yhat != y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.119\n"
     ]
    }
   ],
   "source": [
    "decode = lambda params: (params[:D * C].reshape((D, C)), params[-C:])\n",
    "z = T.dvector()\n",
    "mu = lambda X, W, b: T.nnet.softmax(T.dot(X, W) + b)\n",
    "loss = T.sum(T.nnet.categorical_crossentropy(mu(X, *decode(z)), Y))\n",
    "grad = theano.function([z], T.grad(loss, z))\n",
    "hess = theano.function([z], T.hessian(loss, z))    \n",
    "loss = theano.function([z], loss)\n",
    "\n",
    "params = [0] * (D * C + C)\n",
    "decode = lambda params: (np.array((list(params[:D * C]))).reshape((D, C)), params[-C:])\n",
    "\n",
    "# params = minimize(loss, params, method = 'Newton-CG', jac = grad, hess = hess).x\n",
    "params = minimize(loss, params).x\n",
    "\n",
    "decode = lambda params: (np.array((list(params[:D * C]))).reshape((D, C)), params[-C:])\n",
    "mu = lambda X, W, b: softmax(X.dot(W) + b)\n",
    "yhat = np.argmax(mu(X, *decode(params)), axis = 1)\n",
    "print '%0.3f' % np.mean(yhat != y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.348\n"
     ]
    }
   ],
   "source": [
    "W = theano.shared(np.zeros((D, C)))\n",
    "b = theano.shared(np.zeros(C))\n",
    "tx = T.dmatrix()\n",
    "ty = T.dmatrix()\n",
    "\n",
    "loss = T.sum(T.nnet.categorical_crossentropy(T.nnet.softmax(T.dot(tx, W) + b), ty))\n",
    "learning_rate = 0.01\n",
    "optimizer = theano.function([tx, ty], updates =\n",
    "    [[p, p - T.grad(loss, p) * learning_rate] for p in [W, b]]\n",
    ")\n",
    "\n",
    "optimizer(X, Y)\n",
    "W, b = W.get_value(), b.get_value()\n",
    "yhat = np.argmax(softmax(X.dot(W) + b), axis = 1)\n",
    "print '%0.3f' % np.mean(yhat != y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.348\n"
     ]
    }
   ],
   "source": [
    "W = tf.Variable(tf.zeros((D, C)))\n",
    "b = tf.Variable(tf.zeros(C))\n",
    "\n",
    "loss = tf.nn.softmax_cross_entropy_with_logits(tf.matmul(X, W) + b, Y)\n",
    "optimizer = tf.train.AdamOptimizer().minimize(loss)\n",
    "\n",
    "with tf.Session() as sess:\n",
    "    sess.run(tf.initialize_all_variables())\n",
    "    sess.run(optimizer)\n",
    "    W, b = W.eval(), b.eval()\n",
    "\n",
    "yhat = np.argmax(softmax(X.dot(W) + b), axis = 1)\n",
    "print '%0.3f' % np.mean(yhat != y)"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
